clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java

```diff
@@ -616,6 +620,80 @@ public void testCommitSyncTriggersFencedExceptionFromCommitAsync() {
+        assertThrows(TimeoutException.class, () -> consumer.commitSync(Collections.emptyMap(), Duration.ofMillis(100)));
+
+        // Complete async commit event
+        final ArgumentCaptor<AsyncCommitEvent> commitEventCaptor = ArgumentCaptor.forClass(AsyncCommitEvent.class);
+        verify(applicationEventHandler).add(commitEventCaptor.capture());
+        final AsyncCommitEvent commitEvent = commitEventCaptor.getValue();
+        commitEvent.future().complete(null);
+
+        // Commit async is completed, so commit sync completes immediately (since offsets are empty)
+        assertDoesNotThrow(() -> consumer.commitSync(Collections.emptyMap(), Duration.ofMillis(100)));
+    }
+
+    @Test
+    public void testCommitSyncAwaitsCommitAsyncCompletionWithNonEmptyOffsets() {
+        final TopicPartition tp = new TopicPartition("foo", 0);
+        testSyncCommitTimesoutAfterIncompleteAsyncCommit(tp);
+
+        // Complete async commit event and sync commit event
+        final ArgumentCaptor<AsyncCommitEvent> commitEventCaptor = ArgumentCaptor.forClass(AsyncCommitEvent.class);
+        verify(applicationEventHandler).add(commitEventCaptor.capture());
+        final AsyncCommitEvent commitEvent = commitEventCaptor.getValue();
+        commitEvent.future().complete(null);
+
+        // Commit async is completed, so commit sync completes immediately (since offsets are empty)
+        assertDoesNotThrow(() -> consumer.commitSync(Collections.singletonMap(tp, new OffsetAndMetadata(20)), Duration.ofMillis(100)));
+    }
+
+    @Test
+    public void testCommitSyncAwaitsCommitAsyncButDoesNotFail() {
+        final TopicPartition tp = new TopicPartition("foo", 0);
+        testSyncCommitTimesoutAfterIncompleteAsyncCommit(tp);
+
+        // Complete exceptionally async commit event and sync commit event
+        final ArgumentCaptor<AsyncCommitEvent> commitEventCaptor = ArgumentCaptor.forClass(AsyncCommitEvent.class);
+        verify(applicationEventHandler).add(commitEventCaptor.capture());
+        final AsyncCommitEvent commitEvent = commitEventCaptor.getValue();
+        commitEvent.future().completeExceptionally(new KafkaException("Test exception"));
+
+        // Commit async is completed exceptionally, but this will be handled by commit callback - commit sync should not fail.
+        assertDoesNotThrow(() -> consumer.commitSync(Collections.singletonMap(tp, new OffsetAndMetadata(20)), Duration.ofMillis(100)));
+    }
+
+    private void testSyncCommitTimesoutAfterIncompleteAsyncCommit(TopicPartition tp) {
+        time = new MockTime(1);
+        consumer = newConsumer();
+
+        // Commit async (incomplete)
+        doReturn(Fetch.empty()).when(fetchCollector).collectFetch(any(FetchBuffer.class));
+        doReturn(LeaderAndEpoch.noLeaderOrEpoch()).when(metadata).currentLeader(any());
+        consumer.assign(Collections.singleton(tp));
+        consumer.seek(tp, 20);
```

Comment:
Is the `seek()` here to force the partition out of the `INITIALIZED` state in `SubscriptionState`?

Reply:
It's supposed to give the consumer a valid position, otherwise the commit async doesn't do anything. So yeah, I think that would mean forcing out of `INITIALIZING`? Or what did you mean by `INITIALIZED`..

Reply:
Yes, that's what I meant 😉